app/src/heartbeat/heartbeat_storage_desktop.cc

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "app/src/heartbeat/heartbeat_storage_desktop.h"
+
+#include <fstream>
+#include <regex>
+#include <vector>
+
+#include "app/logged_heartbeats_generated.h"
+#include "app/src/filesystem.h"
+#include "app/src/include/firebase/internal/mutex.h"
+
+namespace firebase {
+namespace heartbeat {
+
+using LoggedHeartbeatsFbs =
+    com::google::firebase::cpp::heartbeat::LoggedHeartbeats;
+using com::google::firebase::cpp::heartbeat::CreateLoggedHeartbeats;
+using com::google::firebase::cpp::heartbeat::CreateUserAgentAndDates;
+using com::google::firebase::cpp::heartbeat::GetLoggedHeartbeats;
+using com::google::firebase::cpp::heartbeat::UserAgentAndDates;
+using com::google::firebase::cpp::heartbeat::VerifyLoggedHeartbeatsBuffer;
+
+// Using an anonymous namespace for file mutex and filename
+namespace {
+
+const char kHeartbeatDir[] = "firebase-heartbeat";
+const char kHeartbeatFilenamePrefix[] = "heartbeats-";
+
+// Returns the mutex that protects accesses to the storage file.
+Mutex& FileMutex() {
+  static Mutex* mutex_ = new Mutex(Mutex::kModeNonRecursive);
+  return *mutex_;
+}
+
+std::string GetFilename(std::string app_id, std::string* error) {
+  std::string app_dir =
+      AppDataDir(kHeartbeatDir, /*should_create=*/true, error);
+  if (app_dir.empty()) {
+    return "";
+  }
+
+  // Remove any symbols from app_id that might not be allowed in filenames.
+  auto app_id_without_symbols =
+      std::regex_replace(app_id, std::regex("[/\\?%*:|\"<>.,;=]"), "");
+  // Note: fstream will convert / to \ if needed on windows.
+  return app_dir + "/" + kHeartbeatFilenamePrefix + app_id_without_symbols;
+}
+
+}  // namespace
+
+HeartbeatStorageDesktop::HeartbeatStorageDesktop(std::string app_id)
+    : filename_(GetFilename(app_id, &error_)) {
+  MutexLock lock(FileMutex());
+
+  // Ensure the file exists, otherwise the first attempt to read it would
+  // fail.
+  std::ofstream file(filename_, std::ios_base::app);
+
+  if (!file) {
+    error_ = "Unable to open '" + filename_ + "'.";
+  }
+}
+
+// Size is arbitrary, just making sure that there is a sane limit.
+static const int kMaxBufferSize = 1024 * 500;
+
+bool HeartbeatStorageDesktop::ReadTo(LoggedHeartbeats* heartbeats_output) {
```

Comment:
http://go/cstyle#Inputs_and_Outputs

If this parameter is required, use `LoggedHeartbeats& heartbeats_output`